rfd/0001-kubernetes-manager.md

```diff
@@ -0,0 +1,100 @@
+---
+authors: Vladimir Kochnev (marshall-lee@evilmartians.com)
+state: draft
+---
+
+# RFD 1 - Kubernetes manager
+
+## What
+
+Teleport Kubernetes Manager is a [controller for Kubernetes](https://kubernetes.io/docs/concepts/architecture/controller/) serving Teleport resources defined as [Kubernetes custom resources](https://kubernetes.io/docs/concepts/extend-kubernetes/api-extension/custom-resources/).
+
+## Why
+
+Automated bootstrap of Teleport data is not an easy task, especially when it's running in Kubernetes. One can use `tctl` from the Teleport CLI toolset to bootstrap users/roles but integrate it into something like [Helm charts](https://helm.sh/docs/topics/charts/) quickly becomes a pain. In many scenarios, it's far more convenient to have the Teleport data defined the same way as the Kubernetes resources like Deployments and Services, in a format very similar to Teleport's own format.
+
+## Details
+
+### Implementation
+
+For the task, the [Kubebuilder](https://kubebuilder.io/) framework should be used. It has great code generation capabilities and gives a complete setup to develop & deploy the manager.
+
+### Instance definitions
+
+To write something into Teleport, the manager needs to know what server to connect to and how. So it needs Teleport Auth/Proxy/Tunnel address and credentials to be recognized. 
```

Comment:
Can K8s operator run as a part of a Teleport process (e.g. auth server)?

Reply:
@r0mant Yes it can but why burden the Teleport with an extra responsibility? Any plugin can theoretically be a part of auth server but we make plugins anyway.
In programming it's better to divide than not to divide unless some serious concerns raised. Can you advance your reasoning?

Reply:
It was decided to run the controller as a part of Teleport deployment i.e. as a sidecar process. So neither instance nor secrets will be needed. Controller would be able to generate credentials for itself.